src/main/drivers/serial_uart_stm32f30x.c

```diff
@@ -148,44 +124,188 @@ static uartDevice_t uartHardware[] = {
+
+// XXX What a waste...
+// XXX These could be left inside the serialUARTx() function,
+// XXX but they will eventually go away...
+// XXX ARRAYLEN(uartHardware) is overkill; What was the number (count) of UART serial ports???
+
+static volatile uint8_t rxBuffers[UART_RX_BUFFER_SIZE * ARRAYLEN(uartHardware)];
+static volatile uint8_t txBuffers[UART_TX_BUFFER_SIZE * ARRAYLEN(uartHardware)];
 
 void serialInitHardwareMap(serialPinConfig_t *pSerialPinConfig)
 {
-    // Force initialize only for USART1 (PA10, PA9)
-    UNUSED(pSerialPinConfig);
-    uartHardware[0].rx = uartHardware[0].pinPair[0].rx;
-    uartHardware[0].tx = uartHardware[0].pinPair[0].tx;
-    uartHardwareMap[0] = &uartHardware[0];
+    memset(uartHardwareMap, 0, sizeof(uartHardwareMap));
+
+    for (size_t index = 0 ; index < ARRAYLEN(uartHardware) ; index++) {
+        uartDevice_t *uartDev = &uartHardware[index];
+        for (int pair = 0 ; pair < UART_PINPAIR_COUNT ; pair++) {
+
+            if (uartDev->pinPair[pair].rx == pSerialPinConfig->ioTagRx[index]
+                    && uartDev->pinPair[pair].tx == pSerialPinConfig->ioTagTx[index]) {
+                uartDev->rx = uartDev->pinPair[pair].rx;
+                uartDev->tx = uartDev->pinPair[pair].tx;
+                uartDev->rxBuffer = &rxBuffers[index * UART_RX_BUFFER_SIZE];
+                uartDev->txBuffer = &txBuffers[index * UART_TX_BUFFER_SIZE];
+                uartHardwareMap[index] = uartDev;
+
+                break;
+            }
+        }
+    }
 }
 
-void serialUARTInit(IO_t tx, IO_t rx, portMode_t mode, portOptions_t options, uint8_t af, uint8_t index)
+void usartIrqHandler(uartPort_t *s)
+{
+    uint32_t ISR = s->USARTx->ISR;
+
+    if (!s->rxDMAChannel && (ISR & USART_FLAG_RXNE)) {
+        if (s->port.rxCallback) {
+            s->port.rxCallback(s->USARTx->RDR);
+        } else {
+            s->port.rxBuffer[s->port.rxBufferHead++] = s->USARTx->RDR;
+            if (s->port.rxBufferHead >= s->port.rxBufferSize) {
+                s->port.rxBufferHead = 0;
+            }
+        }
+    }
+
+    if (!s->txDMAChannel && (ISR & USART_FLAG_TXE)) {
+        if (s->port.txBufferTail != s->port.txBufferHead) {
+            USART_SendData(s->USARTx, s->port.txBuffer[s->port.txBufferTail++]);
+            if (s->port.txBufferTail >= s->port.txBufferSize) {
+                s->port.txBufferTail = 0;
+            }
+        } else {
+            USART_ITConfig(s->USARTx, USART_IT_TXE, DISABLE);
+        }
+    }
+
+    if (ISR & USART_FLAG_ORE)
+    {
+        USART_ClearITPendingBit (s->USARTx, USART_IT_ORE);
+    }
+}
+
+#if defined(USE_UART1_TX_DMA) || defined(USE_UART2_TX_DMA) || defined(USE_UART3_TX_DMA)
+static void handleUsartTxDma(dmaChannelDescriptor_t* descriptor)
+{
+    uartPort_t *s = (uartPort_t*)(descriptor->userParam);
+    DMA_CLEAR_FLAG(descriptor, DMA_IT_TCIF);
+    DMA_Cmd(descriptor->channel, DISABLE);
+
+    if (s->port.txBufferHead != s->port.txBufferTail)
```

Comment:
There is PR to fix race conditions in TxDMA handling ... It may be better to base this work on it